stdpopsim/ext/selection.py

```diff
@@ -151,3 +190,195 @@ def __attrs_post_init__(self):
     @classmethod
     def op_id(cls, op):
         return cls.op_types.index(op)
+
+
+def selective_sweep(
+    single_site_id,
+    population_id,
+    mutation_time,
+    selection_coeff,
+    dominance_coeff=0.5,
+    start_time=None,
+    end_time=None,
+    min_freq_at_start=None,
+    min_freq_at_end=None,
+):
+    """
+    Creates a list of SLiM events corresponding to a selective sweep at a
+    single site (see :meth:`Contig.add_single_site`) that may subsequently be
+    passed to the SLiM simulation engine. The sequence of events is:
+
+        (1) At time `mutation_time`, a **neutral** mutation is
+            introduced into population `population_id`.
+
+        (2) At time `start_time`, the mutation becomes beneficial in population
+            `population_id`. If the frequency of the mutation in population
+            `population_id` is below `min_freq_at_start` in this generation, the
+            simulation will restart from `mutation_time`.
+
+        (3) At time `end_time`, the mutation reverts to neutrality in
+            population `population_id`. If the frequency of the mutation in
+            population `population_id` is below `min_freq_at_end` in this
+            generation, the simulation will restart from `mutation_time`.
+
+    If the mutation is lost at any point between `mutation_time` and
+    `end_time`, the simulation will restart from `mutation_time`.
+
+    During the sweep, the fitnesses of individuals in population
+    `population_id` that are homozygous or heterozygous for the beneficial
+    mutation are `1 + selection_coeff` and  `1 + selection_coeff *
+    dominance_coeff`, respectively. Fitnesses of individuals in other
+    populations are unchanged.
+
+    Because the simulation will restart until the allele frequency conditions
+    are satisfied, care must be taken to ensure that these are reasonable. For
+    example, if `start_time` is very close to `mutation_time` and
+    `min_freq_at_start` is substantially greater than zero, the vast majority
+    of proposed allele frequency trajectories will be rejected and the
+    simulation may take an infeasibly long time to complete.
+
+    :param str single_site_id: The string ID of the single site at which to introduce
+        the mutation, see :meth:`Contig.add_single_site`.
+    :param int population_id: The integer ID of the population in which the mutation is
+        introduced and the sweep occurs.
+    :param float mutation_time: The number of generations in the past at which
```

Comment:
it's nice to call these "time ago" eg `mutation_time_ago`; although if other `time` arguments don't we could not, for consistency.

Reply:
Changed to `xxx_generation_ago` -- this isn't consistent with the argument naming convention used by the extended events themselves, but those aren't exposed/documented so I think it's fine.